### PR TITLE
feat: functional test

### PR DESCRIPTION
**Pull request by kuisathaverat**
[JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582) Convert modules to plugins
[JENKINS-64107](https://issues.jenkins-ci.org/browse/JENKINS-64107) Convert SSHD module to a plugin
[JEP](https://github.com/jenkinsci/jep/pull/326) Convert modules to plugins

It adds three functional test to test the SSHD service, the Git Server, and the Jenkins CLI over SSH.

see https://github.com/jenkinsci/jenkins/pull/5049 and https://github.com/jenkinsci/sshd-module/pull/38

### Desired reviewers

@jenkinsci/code-reviewers

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -1,2 +1,18 @@
 // Builds a module using https://github.com/jenkins-infra/pipeline-library
 buildPlugin(configurations: buildPlugin.recommendedConfigurations())
+
+node('linux'){
+  stage('Funcional tests'){
+    env.FUNCIONAL_TESTS_DIR = 'src/test/funcional'
+    retry(3){
```
Comment (kuisathaverat) [inline Jenkinsfile:7]: retry in case downloads failed due to an environmental issue

File: Jenkinsfile [hunk 2]
```diff
@@ -1,2 +1,18 @@
...
+      sh(label: 'Prepare dependencies', script: "make -C ${FUNCIONAL_TESTS_DIR} dependencies")
+      sh(label: 'Start Jenkins container', script: "make -C ${FUNCIONAL_TESTS_DIR} start")
+    }
+    sh(label: 'Run tests', script: "make -C ${FUNCIONAL_TESTS_DIR} test", returnStatus: true)
```
Comment (kuisathaverat) [inline Jenkinsfile:11]: ignore the return code, if there are failed tests the build will be unstable when we grab the JUnit results.

File: src/test/funcional/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[*]
```
Comment (kuisathaverat) [inline src/test/funcional/.editorconfig:1]: linting rules

File: src/test/funcional/Makefile [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+## @help:download-plugin-manager:Download the Jenkins plugin Manager
+.PHONY: download-plugin-manager
+download-plugin-manager:
+	curl -L -o jenkins/jenkins_home/jenkins-plugin-manager.jar https://github.com/jenkinsci/plugin-installation-manager-tool/releases/latest
+
+## @help:dependencies:Download/copy the dependencies to build Docker images.
+.PHONY: dependencies
+dependencies:
+	# TODO change before merge and after https://github.com/jenkinsci/jenkins/pull/5049 and https://github.com/jenkinsci/sshd-module/pull/38
```
Comment (kuisathaverat) [inline src/test/funcional/Makefile:62]: TODO: I have to modify this part with the correct Jenkins version and using the compiled plugin from sources, but we need to merge the other two PRs first

File: src/test/funcional/customization.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (kuisathaverat) [inline src/test/funcional/customization.py:1]: J2 configuration

File: Jenkinsfile [hunk 6]
```diff
@@ -1,2 +1,18 @@
 // Builds a module using https://github.com/jenkins-infra/pipeline-library
 buildPlugin(configurations: buildPlugin.recommendedConfigurations())
+
+node('linux'){
```
Comment (jglick) [inline Jenkinsfile:4]:
```suggestion
node('docker'){
```

?

File: Jenkinsfile [hunk 7]
```diff
@@ -1,2 +1,18 @@
 // Builds a module using https://github.com/jenkins-infra/pipeline-library
 buildPlugin(configurations: buildPlugin.recommendedConfigurations())
```
Comment (jglick) [inline Jenkinsfile:2]:
```suggestion
buildPlugin()
```

Long story.

File: src/test/funcional/Makefile (hunk 4 above)
Comment (jglick) [reply src/test/funcional/Makefile:62]: So should this be marked as draft status until then?

File: src/test/funcional/Makefile [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	export JENKINS_URL="https://ci.jenkins.io/job/Core/job/jenkins/job/PR-5049/22/artifact/org/jenkins-ci/main/jenkins-war/2.270-rc30619.ae60983c246a/jenkins-war-2.270-rc30619.ae60983c246a.war"; \
+	curl -fsSL -o jenkins/war/jenkins.war $${JENKINS_URL};
+	export SSHD_MODULE_URL="https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/modules/sshd/3.0-rc166.9325e5d37325/sshd-3.0-rc166.9325e5d37325.hpi"; \
```
Comment (jglick) [inline src/test/funcional/Makefile:65]: To be useful, we would need to use the build from the current source tree, as two lines below right?

**Review by jglick — COMMENTED** [review]
There is already functional test coverage for SSHD; is all this just a workaround for JENKINS-41827? If so, I wonder if it is worth the added complexity, vs. just doing a one-time manual sanity check.

**Comment by kuisathaverat** [comment]
> There is already functional test coverage for SSHD; is all this just a workaround for JENKINS-41827? If so, I wonder if it is worth the added complexity, vs. just doing a one-time manual sanity check.

yes, it is a workaround for JENKINS-41827

File: src/test/funcional/Makefile (hunk 8 above)
Comment (kuisathaverat) [reply src/test/funcional/Makefile:65]: yes, this target can rid of the do Jenkins war download (the Docker image will have the proper core), and it has to get the compiled plugin from the sources. It is something to change when we merge the other two PRs

Comment (kuisathaverat) [inline src/test/funcional/Makefile:65]:
```suggestion
	export SSHD_MODULE_URL="https://repo.jenkins-ci.org/releases/org/jenkins-ci/modules/sshd/3.0.1/sshd-3.0.1.hpi"; \
```

File: src/test/funcional/Makefile [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+## @help:download-plugin-manager:Download the Jenkins plugin Manager
+.PHONY: download-plugin-manager
+download-plugin-manager:
+	curl -L -o jenkins/jenkins_home/jenkins-plugin-manager.jar https://github.com/jenkinsci/plugin-installation-manager-tool/releases/latest
+
+## @help:dependencies:Download/copy the dependencies to build Docker images.
+.PHONY: dependencies
+dependencies:
+	# TODO change before merge and after https://github.com/jenkinsci/jenkins/pull/5049 and https://github.com/jenkinsci/sshd-module/pull/38
+	export JENKINS_URL="https://ci.jenkins.io/job/Core/job/jenkins/job/PR-5049/22/artifact/org/jenkins-ci/main/jenkins-war/2.270-rc30619.ae60983c246a/jenkins-war-2.270-rc30619.ae60983c246a.war"; \
```
Comment (kuisathaverat) [inline src/test/funcional/Makefile:63]:
```suggestion
	export JENKINS_URL="https://ci.jenkins.io/job/Core/job/jenkins/job/PR-5049/36/artifact/org/jenkins-ci/main/jenkins-war/2.282-rc30922.733c17f384aa/jenkins-war-2.282-rc30922.733c17f384aa.war"; \
```

**Pull request closed without merging**